File: leetcode-java/src/main/java/com/melalex/leetcode/FloodFill.java
package com.melalex.leetcode;

import java.util.Queue;

public class FloodFill {

  public int[][] floodFill(int[][] image, int sr, int sc, int color) {
    var target = image[sr][sc];
    var m = image.length;
    var n = image[0].length;
    Queue<Pos> queue = new java.util.ArrayDeque<>();

    if (target != color) {
      queue.offer(new Pos(sr, sc));
    }

    while (!queue.isEmpty()) {
      var it = queue.poll();
      var i = it.i;
      var j = it.j;

      if (i >= 0 && i < m && j >= 0 && j < n && image[i][j] == target) {
        image[i][j] = color;

        queue.offer(new Pos(i + 1, j));
        queue.offer(new Pos(i, j + 1));
        queue.offer(new Pos(i - 1, j));
        queue.offer(new Pos(i, j - 1));
      }
    }

    return image;
  }

  private record Pos(int i, int j) {
  }
}

File: leetcode-scala/src/main/scala/com/leetcode/FloodFill.scala
package com.leetcode

object FloodFill {

  def floodFill(image: Array[Array[Int]], sr: Int, sc: Int, color: Int): Array[Array[Int]] = {
    val target = image(sr)(sc)
    val m = image.length
    val n = image(0).length

    def fillInternal(i: Int, j: Int): Unit =
      if (i >= 0 && i < m && j >= 0 && j < n && image(i)(j) == target) {
        image(i)(j) = color

        fillInternal(i + 1, j)
        fillInternal(i, j + 1)
        fillInternal(i - 1, j)
        fillInternal(i, j - 1)
      }

    if (target != color) {
      fillInternal(sr, sc)
    }

    image
  }

  def main(args: Array[String]): Unit = {
    println(floodFill(Array(Array(0, 0, 0), Array(0, 0, 0), Array(0, 0, 0)), 0, 0, 0).map(_.mkString).mkString)
  }
}
